### ISSUE TITLE
ZX80/ZX81 virtual device for sld output

### ISSUE THREAD

**Issue by stevexyz** [z00m128/sjasmplus#232]
DeZog is using sld as sjasm best listing. "The SLD data are being exported only for machine code produced within one of the virtual devices". Willing to use it to develop for ZX80/ZX81 seems new device type is needed.

The ZX80/ZX81 are simple computers, and should not be difficult to add them. If someone will guide me to the places to intervene, I can even make a pull request for it in case.

**Comment by ped7g** [comment]
does zxspectrum48 cover these as sort of "superset" of zx80/81, or does the memory map differ enough to make zx48 map invalid for such use? (sorry for entry level question, I know next to nothing about ZX80 and ZX81, I have just seen them like 2-3 times in my life :) ).

**Comment by stevexyz** [comment]
Despite Spectrum has been my first computer, I had now to search to find information about it :)

Anyway, depending what you mean with "memory map", here some differences I can see:
- zx81 has 8k ROM (zx80 4k) while spectrum 16k (all starting at 0)
- if I got it properly zx spectrum video memory is strictly below the 32k; zx81 way to handle video has not this limit even if in not true-hires it is "executing" video memory in the upper 32k to display the data in the lower 32k (it is a bit complex to describe the tricks used for video display but I can elaborate further if needed)
- "standard" zx80 and zx81 have just 1kb of memory (from 0x4000 to 0x4400)
- both have no paging feature with standard expansions

Waiting for answer here I found in the documentation that there is a recent instruction that might even be used: "DEFDEVICE", but despite my try I was always getting error using it. Can you provide an example? This one of my tests:
```
;DEFINE ZX81
DEFDEVICE zx81, $4400, 1
DEVICE zx81, $4400

...
```

```
Pass 1 complete (0 errors)
Pass 2 complete (0 errors)
loader.asm(2): error: Unrecognized instruction: zx81, $4400, 1
loader.asm(3): error: Unrecognized instruction: zx81, $4400
Pass 3 complete
Errors: 2, warnings: 1, compiled: 131 lines, work time: 0.004 seconds
```

**Comment by ped7g** [comment]
So andrivet in DeZog issue comment suggest to use `NOSLOT64K` as it works for him for ZX81, and that sounds OK to me.

About `DEFDEVICE`: one could argue your slot is rather $400 in size, not $4400. So `DEFDEVICE zx81, $400, 1` would create virtual device with 64 slots (including slot 16 from address $4000 to $43FF which is where the HW has those $400 of RAM positioned) and map into every slot the same first and only page of memory (with the total memory of the virtual device being only $400, but mapped 64 times into the 64ki address space). I don't see benefit to doing this except you would be exercising code paths of sjasmplus which are rarely used, probably only by CI [tests](https://github.com/z00m128/sjasmplus/blob/master/tests/devices/defdevice/defdevice.asm).

The $4400 definition would create device with four uneven slots ($0000..43FF, 4400..87FF, 8800..CBFF, CC00..FFFF), with single page $4400 long mapped to each of these (well, last slot would map only beginning of it because it's only $3400 bytes long), so technically this would work too.

Anyway, I don't see enough benefit in adding specific ZX80/ZX81 device, as there is currently no extra functionality and the general `NOSLOT64K` device covers this use case fully. If there will be need for some specific ZX80/81 functionality, then it may get added, but right now it would be just another alias for block of memory with no extra functionality.

**Comment by maziac** [comment]
Yes, I'm also using the NOSLOT64K myself currently. It works perfectly. The zx81 has a few expansion boards around but as  no banking is done (in the sense that different banks can be switched in for the same slot) the NOSLOT64K will do.

**Comment by maziac** [comment]
The slots/banking can be done with NOSLOT64.
But there is something else to the virtual device I have not considered yet.
It would be very helpful If sjasmplus could save into the .p file format. Just Like sna for Spectrum.
That would make it very easy to create a file that can be read from any emulator.
What do you think?

**Comment by ped7g** [comment]
Is there some nice tech reference docs about binary form of .p? I never looked into zx80/81 stuff, so I have no idea.

As usual with extensions for platform which I'm not familiar, these require someone "sponsoring" the change by providing tech docs, providing some test material, validating end results and answer any ambiguous questions I may have during implementation. And patience to be around when I finally get to it. If you have some docs and example of src -> .p, feel free to post it here and reopen.

**Comment by maziac** [comment]
I could help with that.
The p file is basically just a binary copy from Ram 0x4009 upwards.
What is a bit more complicated is that it saves some system vars as well and the basic program.
We would need a minimal basic program with a REM line that holds our sjasmplus compiled machine code and a USR that jumps to the start of the machine code.
I could provide some code for it. (But give me some time)

Here a ref to the p file: https://k1.spdns.de/Develop/Projects/zasm/Info/O80%20and%20P81%20Format.txt

**Comment by stevexyz** [comment]
> So andrivet in DeZog issue comment suggest to use `NOSLOT64K` as it works for him for ZX81, and that sounds OK to me.

NOSLOT64K is fine, but if a ZX81 synonim could not be an option, I would at least mention it in the documentation
